File: Baasket.LogisticsV2/CheckoutSession.cs
namespace Baasket.LogisticsV2;

/// <summary>
/// Subject in the Observer pattern for a single checkout lifecycle.
/// Mirrors the existing Baasket domain (buyer, payment reference, totals, optional offer)
/// without coupling to the legacy Python <c>logistics.py</c> module.
/// </summary>
public sealed class CheckoutSession : ISubject
{
    private readonly List<IObserver> _observers = [];

    public CheckoutSession(
        Guid sessionId,
        int buyerId,
        string buyerName,
        string buyerEmail,
        string paymentMethod,
        decimal subtotal,
        decimal fee,
        decimal total,
        string paymentReference,
        int? offerId = null,
        int? orderId = null)
    {
        if (buyerId <= 0)
            throw new ArgumentOutOfRangeException(nameof(buyerId));

        SessionId = sessionId;
        BuyerId = buyerId;
        BuyerName = buyerName ?? throw new ArgumentNullException(nameof(buyerName));
        BuyerEmail = buyerEmail ?? throw new ArgumentNullException(nameof(buyerEmail));
        PaymentMethod = paymentMethod ?? throw new ArgumentNullException(nameof(paymentMethod));
        Subtotal = subtotal;
        Fee = fee;
        Total = total;
        PaymentReference = paymentReference ?? throw new ArgumentNullException(nameof(paymentReference));
        OfferId = offerId;
        OrderId = orderId;
        State = CheckoutSessionState.CheckoutStarted;
        CreatedAtUtc = DateTime.UtcNow;
    }

    public Guid SessionId { get; }
    public int BuyerId { get; }
    public string BuyerName { get; }
    public string BuyerEmail { get; }
    public string PaymentMethod { get; }
    public decimal Subtotal { get; }
    public decimal Fee { get; }
    public decimal Total { get; }
    public string PaymentReference { get; }
    public int? OfferId { get; }
    public int? OrderId { get; private set; }
    public CheckoutSessionState State { get; private set; }
    public DateTime CreatedAtUtc { get; }
    public DateTime? PaymentConfirmedAtUtc { get; private set; }
    public DateTime? TrackingUpdatedAtUtc { get; private set; }
    public CheckoutReceipt? Receipt { get; private set; }

    public string StatusDisplay => CheckoutSessionStateDisplay.Display(State);

    public void Attach(IObserver observer)
    {
        ArgumentNullException.ThrowIfNull(observer);

        if (!_observers.Contains(observer))
            _observers.Add(observer);
    }

    public void Detach(IObserver observer)
    {
        ArgumentNullException.ThrowIfNull(observer);
        _observers.Remove(observer);
    }

    public void Notify(CheckoutSessionState previousState)
    {
        // Snapshot so observers can detach safely during notification.
        foreach (var observer in _observers.ToArray())
            observer.Update(this, previousState);
    }

    /// <summary>
    /// Transitions from <see cref="CheckoutSessionState.CheckoutStarted"/> to
    /// <see cref="CheckoutSessionState.PaymentConfirmed"/> and notifies observers.
    /// </summary>
    public void ConfirmPayment(int? orderId = null)
    {
        if (State != CheckoutSessionState.CheckoutStarted)
            throw new InvalidOperationException(
                $"Payment can only be confirmed from {CheckoutSessionState.CheckoutStarted}. Current state: {State}.");

        var previousState = State;
        State = CheckoutSessionState.PaymentConfirmed;
        PaymentConfirmedAtUtc = DateTime.UtcNow;

        if (orderId.HasValue)
            OrderId = orderId;

        Notify(previousState);
    }

    /// <summary>
    /// Called by <see cref="ReceiptGeneratorObserver"/> after payment confirmation.
    /// </summary>
    public void RecordReceipt(CheckoutReceipt receipt)
    {
        ArgumentNullException.ThrowIfNull(receipt);

        if (State != CheckoutSessionState.PaymentConfirmed)
            throw new InvalidOperationException("Receipt can only be recorded after payment is confirmed.");

        if (Receipt is not null)
            throw new InvalidOperationException("Receipt has already been generated for this session.");

        Receipt = receipt;
    }

    /// <summary>
    /// Advances tracking one step forward (Packed → Shipped → Delivered).
    /// Called by <see cref="AutomatedTrackingObserver"/> timers.
    /// </summary>
    public void AdvanceTo(CheckoutSessionState nextState)
    {
        if (!IsValidForwardTransition(State, nextState))
            throw new InvalidOperationException(
                $"Invalid transition: {State} → {nextState}.");

        var previousState = State;
        State = nextState;
        TrackingUpdatedAtUtc = DateTime.UtcNow;
        Notify(previousState);
    }

    private static bool IsValidForwardTransition(
        CheckoutSessionState current,
        CheckoutSessionState next)
    {
        return (current, next) switch
        {
            (CheckoutSessionState.PaymentConfirmed, CheckoutSessionState.Packed) => true,
            (CheckoutSessionState.Packed, CheckoutSessionState.Shipped) => true,
            (CheckoutSessionState.Shipped, CheckoutSessionState.Delivered) => true,
            _ => false,
        };
    }
}

File: logistics_v2/checkout_session.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from logistics_v2.checkout_receipt import CheckoutReceipt
from logistics_v2.checkout_session_state import CheckoutSessionState, display
from logistics_v2.interfaces import IObserver


class CheckoutSession:
    """Subject for the LogisticsV2 checkout lifecycle."""

    _FORWARD_TRANSITIONS = {
        (CheckoutSessionState.PAYMENT_CONFIRMED, CheckoutSessionState.PACKED),
        (CheckoutSessionState.PACKED, CheckoutSessionState.SHIPPED),
        (CheckoutSessionState.SHIPPED, CheckoutSessionState.DELIVERED),
    }

    def __init__(
        self,
        *,
        session_id: UUID,
        buyer_id: int,
        buyer_name: str,
        buyer_email: str,
        payment_method: str,
        subtotal: Decimal,
        fee: Decimal,
        total: Decimal,
        payment_reference: str,
        offer_id: int | None = None,
        shipping_id: int | None = None,
    ) -> None:
        if buyer_id <= 0:
            raise ValueError("buyer_id must be positive")

        self.session_id = session_id
        self.buyer_id = buyer_id
        self.buyer_name = buyer_name
        self.buyer_email = buyer_email
        self.payment_method = payment_method
        self.subtotal = subtotal
        self.fee = fee
        self.total = total
        self.payment_reference = payment_reference
        self.offer_id = offer_id
        self.shipping_id = shipping_id
        self.state = CheckoutSessionState.CHECKOUT_STARTED
        self.created_at_utc = datetime.utcnow()
        self.payment_confirmed_at_utc: datetime | None = None
        self.tracking_updated_at_utc: datetime | None = None
        self.receipt: CheckoutReceipt | None = None
        self._observers: list[IObserver] = []

    @property
    def status_display(self) -> str:
        return display(self.state)

    def attach(self, observer: IObserver) -> None:
        if observer not in self._observers:
            self._observers.append(observer)

    def detach(self, observer: IObserver) -> None:
        self._observers.remove(observer)

    def notify(self, previous_state: CheckoutSessionState) -> None:
        for observer in list(self._observers):
            observer.update(self, previous_state)

    def confirm_payment(self, shipping_id: int | None = None) -> None:
        if self.state is not CheckoutSessionState.CHECKOUT_STARTED:
            raise RuntimeError(
                f"Payment can only be confirmed from {CheckoutSessionState.CHECKOUT_STARTED}. "
                f"Current state: {self.state}."
            )

        previous_state = self.state
        self.state = CheckoutSessionState.PAYMENT_CONFIRMED
        self.payment_confirmed_at_utc = datetime.utcnow()
        if shipping_id is not None:
            self.shipping_id = shipping_id
        self.notify(previous_state)

    def record_receipt(self, receipt: CheckoutReceipt) -> None:
        if self.state is not CheckoutSessionState.PAYMENT_CONFIRMED:
            raise RuntimeError("Receipt can only be recorded after payment is confirmed.")
        if self.receipt is not None:
            raise RuntimeError("Receipt has already been generated for this session.")
        self.receipt = receipt

    def advance_to(self, next_state: CheckoutSessionState) -> None:
        if (self.state, next_state) not in self._FORWARD_TRANSITIONS:
            raise RuntimeError(f"Invalid transition: {self.state} → {next_state}.")

        previous_state = self.state
        self.state = next_state
        self.tracking_updated_at_utc = datetime.utcnow()
        self.notify(previous_state)
